metadata/src/main/java/org/apache/kafka/metadata/migration/KRaftMigrationZkWriter.java

```diff
@@ -184,11 +232,36 @@ void handleTopicsDelta(Function<Uuid, String> deletedTopicNameResolver, TopicsDe
                         topicDelta.partitionChanges(),
                         migrationState));
             } else {
-                operationConsumer.accept(
-                    "Updating Partitions for Topic " + topicDelta.name() + ", ID " + topicId,
-                    migrationState -> migrationClient.topicClient().updateTopicPartitions(
-                        Collections.singletonMap(topicDelta.name(), topicDelta.partitionChanges()),
-                        migrationState));
+                if (topicDelta.hasPartitionsWithAssignmentChanges())
+                    operationConsumer.accept(
+                        "Updating Topic " + topicDelta.name() + ", ID " + topicId,
+                        migrationState -> migrationClient.topicClient().updateTopic(
+                            topicDelta.name(),
+                            topicId,
+                            topicsImage.getTopic(topicId).partitions(),
+                            migrationState));
+                Map<Integer, PartitionRegistration> newPartitions = topicDelta.newPartitions();
+                Map<Integer, PartitionRegistration> changedPartitions = topicDelta.partitionChanges();
+                if (!newPartitions.isEmpty()) {
+                    operationConsumer.accept(
+                        "Create new partitions for Topic " + topicDelta.name() + ", ID " + topicId,
+                        migrationState -> migrationClient.topicClient().createTopicPartitions(
+                            Collections.singletonMap(topicDelta.name(), topicDelta.partitionChanges()),
+                            migrationState));
+                    changedPartitions = changedPartitions
```

Comment:
Instead of reassigning, could we just remove the newPartitions from changedPartitions?